webrender/src/frame.rs

```diff
@@ -66,45 +54,21 @@ pub struct Frame {
     frame_builder: Option<FrameBuilder>,
 }
 
-enum SceneItemKind<'a> {
-    StackingContext(&'a SceneStackingContext, PipelineId),
-    Pipeline(&'a ScenePipeline)
+trait DisplayListHelpers {
+    fn starting_stacking_context<'a>(&'a self) -> Option<&'a StackingContext>;
 }
 
-#[derive(Clone)]
-struct SceneItemWithZOrder {
-    item: SceneItem,
-    z_index: i32,
-}
-
-impl<'a> SceneItemKind<'a> {
-    fn collect_scene_items(&self, scene: &Scene) -> Vec<SceneItem> {
-        let mut result = Vec::new();
-        let stacking_context = match *self {
-            SceneItemKind::StackingContext(stacking_context, _) => {
-                &stacking_context.stacking_context
-            }
-            SceneItemKind::Pipeline(pipeline) => {
-                if let Some(background_draw_list) = pipeline.background_draw_list {
-                    result.push(SceneItem {
-                        specific: SpecificSceneItem::DrawList(background_draw_list),
-                    });
-                }
-
-                &scene.stacking_context_map
-                      .get(&pipeline.root_stacking_context_id)
-                      .unwrap()
-                      .stacking_context
-            }
+impl DisplayListHelpers for Vec<DisplayItem> {
+    fn starting_stacking_context<'a>(&'a self) -> Option<&'a StackingContext> {
+        let first_item = match self.first() {
+            Some(item) => item,
+            None => return None,
         };
 
-        for display_list_id in &stacking_context.display_lists {
-            let display_list = &scene.display_list_map[display_list_id];
-            for item in &display_list.items {
-                result.push(item.clone());
-            }
+        match first_item.item {
```

Comment:
That's really minor and doesn't need to be fixed! Could be done in a single match (like I specified in another comment), in nested matches, or even with `and_then` construct:
```
self.first().and_then(|item| match item.item {...})
```